### ISSUE TITLE
Fullscreen cropped when scale factor changes

### ISSUE THREAD

**Issue by ghuser404** [glfw/glfw#1582]
This is on the latest Windows 10. Probably a Windows bug that will get fixed sometime in future, but I'll just leave this here, since GLFW is affected.

1. Make sure monitor scale factor is at 100%
2. Create window in windowed mode (can use Boing example)
3. Set monitor scale factor to 125%
4. Set window mode to fullscreen (Alt+Enter for Boing)
5. Now Alt+Tab and change monitor scale factor to 100%
6. Open your window - it now doesn't cover the screen and appears cropped

**Comment by elmindreda** [comment]
This happens on a Windows Insider build? I couldn't reproduce it reproduce on 1903.

**Comment by ghuser404** [comment]
That's on 1903. Maybe I have some bugged dll left over... such things happened to me in the past, not related to scaling. I may be able to try this on another pc.

**Comment by elmindreda** [comment]
Yup, or it could be my PC having a non-standard setting somewhere. Either way, thank you for finding and reporting this, as always! Hope we can pin it down.

**Comment by ghuser404** [comment]
Tried this on other PC, the same problem.

**Comment by ghuser404** [comment]
Someone has had the same issue (posted 2 years ago) https://stackoverflow.com/questions/47019924/windows-10-dpi-scaling-and-fullscreen

>After further testing, the WM_GETDPISCALEDSIZE message doesn't even seem to be sent if the window is fullscreen, only if it's windowed (WS_POPUP works but only if the size isn't the same as the monitor). Therefore I can't even override this behaviour, and Windows 8.1 works different to Windows 10 v1703 for fullscreened windows.

**Comment by ghuser404** [comment]
@elmindreda This is what I've found in Micirosoft's Terminal repo https://github.com/microsoft/terminal/blob/master/src/interactivity/win32/windowproc.cpp

I believe, if even Microsoft thinks DPICHANGED should be ignored for fullscreen windows, then GLFW should do the same. However, a mere `if (window->monitor)` check will not be enough - this should be done for borderless monitor-sized windows as well, as those behave the same way on Windows -  and I'm not sure how this could be done on the GLFW side.
>// If we're a full screen window, completely ignore what the DPICHANGED says as it will be bigger than the monitor and
// instead just ensure that the window is still taking up the full screen.

```
case WM_DPICHANGED:
{
    _fInDPIChange = true;
    ServiceLocator::LocateGlobals().dpi = HIWORD(wParam);
    _UpdateSystemMetrics();
    s_ReinitializeFontsForDPIChange();

    if (IsInFullscreen())
    {
        // If we're a full screen window, completely ignore what the DPICHANGED says as it will be bigger than the monitor and
        // instead just ensure that the window is still taking up the full screen.
        SetIsFullscreen(true);
    }
    else
    {
        // this is the RECT that the system suggests.
        RECT* const prcNewScale = (RECT*)lParam;
        SetWindowPos(hWnd, HWND_TOP, prcNewScale->left, prcNewScale->top, RECT_WIDTH(prcNewScale), RECT_HEIGHT(prcNewScale), SWP_NOZORDER | SWP_NOACTIVATE);
    }

    _fInDPIChange = false;

    break;
}
```

**Comment by ghuser404** [comment]
@elmindreda I found solution, it works very well. Could you try to reproduce this issue again (maybe you missed something in my description) and see if this solution is good for GLFW?
```
RECT window_area, monitor_area;
GetWindowRect(window->win32.handle, &window_area);
GetWindowRect(GetDesktopWindow(), &monitor_area);

if (!EqualRect(&window_area, &monitor_area))
    ...
```
```
case WM_DPICHANGED:
{
    const float xscale = HIWORD(wParam) / (float) USER_DEFAULT_SCREEN_DPI;
    const float yscale = LOWORD(wParam) / (float) USER_DEFAULT_SCREEN_DPI;

    // Only apply the suggested size if the OS is new enough to have
    // sent a WM_GETDPISCALEDSIZE before this
    if (_glfwIsWindows10CreatorsUpdateOrGreaterWin32())
    {
        RECT window_area, monitor_area;
        GetWindowRect(window->win32.handle, &window_area);
        GetWindowRect(GetDesktopWindow(), &monitor_area);

        if (!EqualRect(&window_area, &monitor_area))
        {
            RECT* suggested = (RECT*) lParam;
            SetWindowPos(window->win32.handle, HWND_TOP,
                         suggested->left,
                         suggested->top,
                         suggested->right - suggested->left,
                         suggested->bottom - suggested->top,
                         SWP_NOACTIVATE | SWP_NOZORDER);
        }
    }

    _glfwInputWindowContentScale(window, xscale, yscale);
    break;
}
```

**Comment by elmindreda** [comment]
@ghuser404 Thank you for all that information! I should be able to reproduce the bug now.

**Comment by ghuser404** [comment]
@elmindreda Thanks for the fix, it works well for the fullscreen scenario and all. However, this doesn't fix it for the borderless window that covers the monitor (aka fullscreen borderless mode) - Windows has special treatment for such windows.

I know the original ticket description didn't specify this scenario. Should we reopen this ticket or create a new one?
